### PR TITLE
MAINT: remove deprecation warning for **kwargs

### PR DESCRIPTION
**Pull request by adriangb**
This PR removes the deprecation warning for **kwargs, hence making it a feature that will stay.

### CODE REVIEW

**Comment by adriangb** [comment]
ccing @stsievert

**Comment by stsievert** [comment]
Hm... I think this is a judgement call.

I think having Keras compatibility is a core feature of this library. But this library also tries to bring the Scikit-learn API to Keras models, and the Scikit-learn API says `fit`/etc should only accept data-dependent arguments (not `batch_size`/etc).

If I were developing this library, I think I'd keep an error message but remove all specific version info (and maybe reword the message a bit to warn about the API difference).

**Comment by adriangb** [comment]
Let me try some rewording instead of removal and let's see how it looks.

**Comment by adriangb** [comment]
Here's the new proposed warning:
```
Use of `**kwargs` for `fit` is not fully supported by the sklearn API.  Instead, set fit arguments at initialization (i.e., `BaseWrapper(fit__batch_size=32)`)
```

I also added a `break` so that only 1 kwarg is warned against each time `fit`/`predict` is called.

**Comment by stsievert** [comment]
> I also added a break so that only 1 kwarg is warned against each time fit/predict is called.

Eh, the goal of this warning is to encourage changing behavior. Making the warnings extra annoying for multiple arguments is a good way to do that. Or, you could consolidate the message:

> Only keyword arguments that depend on the data `X` and `y` should be passed to `fit`. The keyword arguments `batch_size`, `verbose` should be set via the parameters `fit__batch_size`/`batch_size`, `fit__verbose`/`verbose`. This can be done via `set_params` or at initialization with `BaseWrapper(..., fit__batch_size=32, fit__verbose=False)`.

**Comment by adriangb** [comment]
Point noted, thanks for the feedback. Let me try another revision.

File: docs/source/migration.rst [hunk 1]
```diff
@@ -70,7 +67,18 @@ Or to declare separate values for ``fit`` and ``predict``:
...
+
+If you want to change the parameters on a live instance, you can do:
+
+.. code:: python
+
+   clf = KerasClassifier(...)
+   clf.set_params(fit__batch_size=32, predict__batch_size=10000)
+   clf.fit(...)
+
+This effectively gives you the same result as passing ``**kwargs``, just with
```
Comment (stsievert) [inline docs/source/migration.rst:80]:
```suggestion
Functionally, this is the same as passing these parameters during initialization, just with
```

File: scikeras/wrappers.py [hunk 2]
```diff
@@ -35,7 +35,11 @@
 from scikeras.utils.transformers import ClassifierLabelEncoder, RegressorTargetEncoder
 
 
-_kwarg_warn = "Passing `**kwargs` to `{0}` is not fully supported by the sklearn API. Instead, set arguments at initialization (`BaseWrapper({0}__{1}={2})`) or use `set_params`"
+_kwarg_warn = """Passing `**kwargs` to `{0}` is not fully supported by the sklearn API.
```
Comment (stsievert) [inline scikeras/wrappers.py:38]:
```suggestion
_kwarg_warn = """Passing parameters during to `{0}` is not supported by the Scikit-learn API.
```

File: scikeras/wrappers.py [hunk 3]
```diff
@@ -35,7 +35,11 @@
...
+See our docs for more details: https://www.adriangb.com/scikeras/refs/heads/master/migration.html#variable-keyword-arguments-in-fit-and-predict
+The following `**kwargs` were used:
```
Comment (stsievert) [inline scikeras/wrappers.py:40]:
```suggestion
The following parameters were passed to `{0}`:
```

File: scikeras/wrappers.py [hunk 4]
```diff
@@ -706,9 +710,10 @@ def fit(self, X, y, sample_weight=None, **kwargs) -> "BaseWrapper":
             A reference to the instance that can be chain called
             (ex: instance.fit(X,y).transform(X) )
         """
-        for k, v in kwargs.items():
-            warnings.warn(_kwarg_warn.format("fit", k, v))
-            break  # only warn about the first kwarg
+        if kwargs:
+            example_kwarg = next(iter(kwargs.keys()))
+            kwarg_list = "`\n * `".join(kwargs.keys()) + "`"
```
Comment (stsievert) [inline scikeras/wrappers.py:715]: Shouldn't the error message display the value too?

File: scikeras/wrappers.py (hunk 3 above)
Comment (stsievert) [reply scikeras/wrappers.py:40]: And shouldn't the error message provide a method to resolve the warning?

File: docs/source/migration.rst (hunk 1 above)
Comment (adriangb) [reply docs/source/migration.rst:80]: I am specifically trying to give users a way that they can do this if they do _not_ want to pass the parameters via initialization, hence the comparison to passing `**kwargs`. The first part looks good though.

File: scikeras/wrappers.py (hunk 4 above)
Comment (adriangb) [reply scikeras/wrappers.py:715]: What are the values needed for? I don't see how a developer would use that to fix the issue, they can just look up the value in their code.

File: scikeras/wrappers.py (hunk 3 above)
Comment (adriangb) [reply scikeras/wrappers.py:40]: It links to the docs, which I think provide a more detailed explanation and avenues for resolution than is possible in a warning. Do you feel that linking to the docs does not accomplish this?

File: scikeras/wrappers.py (hunk 2 above)
Comment (adriangb) [reply scikeras/wrappers.py:38]: `Passing parameters during to`?  Is that an extra `to`? Wouldn't this be confused with `fit takes no parameters`?

File: scikeras/wrappers.py (hunk 4 above)
Comment (stsievert) [reply scikeras/wrappers.py:715]: Why not show the values? There's no real downside, especially because you're showing each value on a newline at most initialization parameters are basic types (int, str, float, etc).

File: scikeras/wrappers.py (hunk 3 above)
Comment (stsievert) [reply scikeras/wrappers.py:40]: Yes, linking to the docs provides an example. Yes, users can look at the docs and that will resolve their issue. In fact, you *should* link to the full example where you resolve the issue.

I'm not asking to remove the link. I think an error message should say "to resolve this issue, [method]. For more detail, see [link]."

I think this makes SciKeras easier to use because it requires less work to get the code to run. What if I'm developing without internet access? I've often done that on airplanes or road trips.

File: scikeras/wrappers.py (hunk 2 above)
Comment (stsievert) [reply scikeras/wrappers.py:38]: I meant "parameters" as in "estimator parameters" (e.g., `batch_size`). I've modified the suggestion.

File: scikeras/wrappers.py (hunk 3 above)
Comment (adriangb) [reply scikeras/wrappers.py:40]: That's fair. Do you think we can say `put the parameters in the constructor or use set_params` or something simple like that, without injecting parameter-value examples?

File: docs/source/migration.rst (hunk 1 above)
Comment (stsievert) [reply docs/source/migration.rst:80]: I'm confused by your statement. I only tried to replace "`**kwargs`" with words. Wouldn't you pass `**kwargs` to `__init__`?

File: scikeras/wrappers.py (hunk 4 above)
Comment (adriangb) [reply scikeras/wrappers.py:715]: > There's no real downside

I think the main downside for long/large warnings is that the information gets lost in the noise. Repeated 20+ line warnings can make it hard to know what is even important.

> most initialization parameters are basic types

You are right, most are basic types. But several, like loss, optimizer, metrics, callbacks, etc. are at the very least iterables/dicts and at the very worst dicts of instances.

File: docs/source/migration.rst (hunk 1 above)
Comment (adriangb) [reply docs/source/migration.rst:80]: Sorry if it was confusing. What I meant is that this should be the same as passing `**kwargs` to `fit`. So combining my thoughts and yours, I think it would look something like:
```
Functionally, this is the same as passing these parameters to ``fit``, just with one more function call.
```

File: scikeras/wrappers.py (hunk 4 above)
Comment (stsievert) [reply scikeras/wrappers.py:715]: > Repeated 20+ line warnings can make it hard to know what is even important.

Repetition is not an issue. From the [warnings docs][1], "Repetitions of a particular warning for the same source location are typically suppressed." Plus, this warning is only issued once per `fit`/`predict` call (which aren't repeated very often; I would say `partial_fit` is repeated frequently).

I think having one 20+ line warning is okay. I think this warning will be short (<= 4 lines) the vast majority of times it's issued.


[1]:https://docs.python.org/3/library/warnings.html

File: scikeras/wrappers.py (hunk 3 above)
Comment (stsievert) [reply scikeras/wrappers.py:40]: Why the hesitation to include parameter-value examples? I suppose you could use an illustrative example like `foo="bar"`.

File: scikeras/wrappers.py (hunk 4 above)
Comment (adriangb) [reply scikeras/wrappers.py]: You make some good points. Besides, we are currently printing these messages with the values, etc, so it's certainly not going to be _worse_ than what we currently have.

I've pushed a commit to print the values, as well as give examples using fixed dummy `foo=bar` kwargs (https://github.com/adriangb/scikeras/pull/198#discussion_r577793274).

File: scikeras/wrappers.py [hunk 5]
```diff
@@ -925,7 +926,6 @@ def predict(self, X, **kwargs):
             and n_features is the number of features.
         **kwargs : Dict[str, Any]
             Extra arguments to route to ``Model.predict``.
-            This functionality has been deprecated, and will be removed in SciKeras 1.0.0.
```
[the remark this answers is not in the document]
Comment (stsievert) [reply scikeras/wrappers.py:928]: This goes for all the functions below too.

**Review by stsievert — COMMENTED** [review]
This looks a lot better.

Style nit: here's some rewording I would do around the warning (more lines, reorganizing words for `**kwargs`):

```
_kwarg_warn = """Passing estimator parameters as keyword arguments to `{0}` (aka as `**kwargs`)  is not supported by the Scikit-Learn API, and will be deprecated in SciKeras.

To resolve this issue, either set these parameters in the constructor (`BaseWrapper(..., foo=bar)`)  or via `set_params` (`est.set_params(foo=bar)`). The following parameters were passed to `{0}`:

{1}

More detail is available at https://www.adriangb.com/scikeras/migration.html#variable-keyword-arguments-in-fit-and-predict
"""
```

File: scikeras/wrappers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adriangb) [reply scikeras/wrappers.py:928]: warning section added

**Comment by adriangb** [comment]
> Style nit: here's some rewording I would do around the warning (more lines, reorganizing words for `**kwargs`):

Incorporated for the most part

File: scikeras/wrappers.py [hunk 6]
```diff
@@ -697,20 +704,23 @@ def fit(self, X, y, sample_weight=None, **kwargs) -> "BaseWrapper":
...
         **kwargs : Dict[str, Any]
             Extra arguments to route to ``Model.fit``.
-            This functionality has been deprecated, and will be removed in SciKeras 1.0.0.
-            These parameters can also be specified by prefixing `fit__` to a parameter at initialization;
-            e.g, `BaseWrapper(..., fit__batch_size=32, predict__batch_size=1000)`.
+
+        Warnings
+        --------
+            Passing estimator parameters as keyword arguments (aka as ``**kwargs``) to ``fit`` is not supported by the Scikit-Learn API,
+            and may be deprecated in SciKeras.
```
Comment (stsievert) [inline scikeras/wrappers.py:711]:
```suggestion
            and will be deprecated in SciKeras.
```

A ways down the road, I think this behavior will be deprecated – probably a year or two after TF drops support for Scikit-learn wrappers.

I don't think warnings should be issued for behavior that "may be deprecated."

**Review by stsievert — COMMENTED** [review]
Small style nit: deprecate is means to be "will be removed."

File: scikeras/wrappers.py (hunk 3 above)
Comment (adriangb) [reply scikeras/wrappers.py]: > I suppose you could use an illustrative example like foo="bar"

The warning now has an illustrative example like you suggest above.

File: scikeras/wrappers.py (hunk 2 above)
Comment (adriangb) [reply scikeras/wrappers.py]: This suggestion is now outdated, but we have largerly incorporated it, so I am going to resolve this comment.

**Comment by adriangb** [comment]
@stsievert  could you take a look at the unresolved comments? Thanks

File: docs/source/migration.rst (hunk 1 above)
Comment (stsievert) [reply docs/source/migration.rst]: > Functionally, this is the same as passing these parameters to ``fit``, just with one more function call.

Yeah, I think that's a good representation of the message you were trying to convey. I might add a couple words about how it's better because it follows the Sciikit-learn API.

File: scikeras/wrappers.py [hunk 7]
```diff
@@ -36,12 +35,21 @@
 from scikeras.utils.transformers import ClassifierLabelEncoder, RegressorTargetEncoder
 
 
+_kwarg_warn = """Passing estimator parameters as keyword arguments (aka as `**kwargs`) to `{0}` is not supported by the Scikit-Learn API, and will be removed in a future version of SciKeras.
+
+To resolve this issue, either set these parameters in the constructor (`BaseWrapper(..., foo=bar)`) or via `set_params` (`est.set_params(foo=bar)`). The following parameters were passed to `{0}`:
```
Comment (stsievert) [inline scikeras/wrappers.py:40]:
```suggestion
To resolve this issue, either set these parameters in the constructor (e.g., `est = BaseWrapper(..., foo=bar)`) or via `set_params` (e.g., `est.set_params(foo=bar)`). The following parameters were passed to `{0}`:
```

File: docs/source/migration.rst (hunk 1 above)
Comment (adriangb) [reply docs/source/migration.rst]: Edited to:

```
Functionally, this is the same as passing these parameters to ``fit``, just with one more function call.
This is much more compatible with the Scikit-Learn API.
In fact, this is what Scikti-Learn does in the background for hyperparameter tuning.
```

**Comment by adriangb** [comment]
Thanks for this last review @stsievert . I think this is ready.
